File: impl/ruby/test/finished_kat.rb
# frozen_string_literal: true

# Standards-derived, NON-CIRCULAR known-answer test for the draft-00 Finished verify_data
# (binding spec/10 section 6.2; RFC 8446 4.4.4): verify_data = HMAC(finished_key, transcript_hash)
# under the profile hash (SHA-256 at Standard). Ruby mirror of the Go/Python/TS reference tests
# against the SAME pinned vector (test-vectors/v1/finished-kat.json).
#
# Three legs: ANCHOR (HMAC-SHA-256 reproduces RFC 4231 TC1/TC2), ORACLE (independent OpenSSL::HMAC,
# no compute_finished), IMPL (compute_finished + verify_finished accept/reject). Run:
#   ruby test/finished_kat.rb

require "json"
require "openssl"
require_relative "../lib/npamp"

VEC = File.expand_path("../../../test-vectors/v1", __dir__)
FINISHED_KAT_SHA256 = "25c21b0bd3b3b6b77862f4a819f81ff5e4ff42e4b1d70af81feeedc5aad73c7f"

$failures = 0

def check(name, ok)
  if ok
    puts "PASS #{name}"
  else
    puts "FAIL #{name}"
    $failures += 1
  end
end

def load_kat
  raw = File.read(File.join(VEC, "finished-kat.json"), mode: "rb")
  got = OpenSSL::Digest.new("SHA256").hexdigest(raw)
  raise "Finished KAT vector SHA-256 mismatch (swapped vector?): #{got}" unless got == FINISHED_KAT_SHA256

  JSON.parse(raw)
end

def hx(str)
  [str].pack("H*").b
end

# Standard HMAC-SHA-256, independent of compute_finished.
def hmac_oracle(key, data)
  OpenSSL::HMAC.digest("SHA256", key, data)
end

def test_rfc4231_anchor
  kat = load_kat
  [["TC1", kat["rfc4231_hmac_sha256"]["tc1"]], ["TC2", kat["rfc4231_hmac_sha256"]["tc2"]]].each do |label, tc|
    got = Npamp.to_hex(hmac_oracle(hx(tc["key"]), hx(tc["data"])))
    check("finished_anchor_rfc4231_#{label}", got == tc["hmac_sha256"])
  end
end

def test_oracle
  kat = load_kat
  nn = kat["npamp_inputs"]
  exp = kat["expected"]
  check("finished_oracle_server",
        Npamp.to_hex(hmac_oracle(hx(nn["finished_key_server"]), hx(nn["th_scv"]))) == exp["verify_data_server"])
  check("finished_oracle_client",
        Npamp.to_hex(hmac_oracle(hx(nn["finished_key_client"]), hx(nn["th_ccv"]))) == exp["verify_data_client"])
end

def test_impl
  kat = load_kat
  nn = kat["npamp_inputs"]
  exp = kat["expected"]
  [
    ["server", nn["finished_key_server"], nn["th_scv"], exp["verify_data_server"]],
    ["client", nn["finished_key_client"], nn["th_ccv"], exp["verify_data_client"]]
  ].each do |name, fk, th, want|
    fkb = hx(fk)
    thb = hx(th)
    wantb = hx(want)
    check("finished_impl_#{name}_compute", Npamp.to_hex(Npamp.compute_finished(fkb, thb, true)) == want)
    check("finished_impl_#{name}_verify_accept", Npamp.verify_finished(fkb, thb, wantb, true))
    bad = wantb.dup
    bad.setbyte(0, bad.getbyte(0) ^ 0x01)
    check("finished_impl_#{name}_verify_reject_tamper", !Npamp.verify_finished(fkb, thb, bad, true))
  end
end

test_rfc4231_anchor
test_oracle
test_impl

puts($failures.zero? ? "ALL PASS" : "FAILURES: #{$failures}")
exit($failures.zero? ? 0 : 1)

File: impl/php/test/finished_kat.php
<?php

/**
 * Standards-derived, NON-CIRCULAR known-answer test for the draft-00 Finished
 * verify_data (binding spec/10 section 6.2; RFC 8446 section 4.4.4):
 * verify_data = HMAC(finished_key, transcript_hash) under the profile hash
 * (SHA-256 at Standard). PHP mirror of the Go/Python/TS reference tests against
 * the SAME pinned vector (test-vectors/v1/finished-kat.json).
 *
 * Three legs:
 *   ANCHOR - HMAC-SHA-256 reproduces RFC 4231 TC1/TC2.
 *   ORACLE - an independent hash_hmac (no computeFinished) reproduces verify_data,
 *            guarding the vector.
 *   IMPL   - computeFinished reproduces verify_data; verifyFinished accepts the
 *            correct MAC and rejects a tampered one, guarding the impl.
 *
 * Run from impl/php:  php test/finished_kat.php
 */

declare(strict_types=1);

require __DIR__ . '/../src/Npamp.php';

use Sh\Bubblefish\Npamp\Handshake;

$failures = 0;

function check(string $name, bool $ok): void
{
    global $failures;
    if ($ok) {
        echo "ok   - {$name}\n";
    } else {
        echo "FAIL - {$name}\n";
        $failures++;
    }
}

const VEC_DIR = __DIR__ . '/../../../test-vectors/v1';
const FINISHED_KAT_SHA256 = '25c21b0bd3b3b6b77862f4a819f81ff5e4ff42e4b1d70af81feeedc5aad73c7f';

/** Load and SHA-256-pin the vector (fail loud on a swapped/corrupt vector). */
function loadKat(): array
{
    $raw = @file_get_contents(VEC_DIR . '/finished-kat.json');
    if ($raw === false) {
        fwrite(STDERR, "cannot read finished-kat.json\n");
        exit(2);
    }
    $got = hash('sha256', $raw);
    if (!hash_equals(FINISHED_KAT_SHA256, $got)) {
        fwrite(STDERR, "Finished KAT vector SHA-256 mismatch (swapped vector?): {$got}\n");
        exit(2);
    }
    return json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
}

function hx(string $s): string
{
    $out = $s === '' ? '' : hex2bin($s);
    if ($out === false) {
        fwrite(STDERR, "bad hex: {$s}\n");
        exit(2);
    }
    return $out;
}

/** Standard HMAC-SHA-256, independent of computeFinished. */
function hmacOracle(string $key, string $data): string
{
    return hash_hmac('sha256', $data, $key, true);
}

$k = loadKat();

// --- ANCHOR: HMAC-SHA-256 reproduces the RFC 4231 TC1/TC2 published MACs. ---
foreach (['tc1', 'tc2'] as $label) {
    $tc = $k['rfc4231_hmac_sha256'][$label];
    $got = bin2hex(hmacOracle(hx($tc['key']), hx($tc['data'])));
    check("finished anchor: HMAC-SHA-256 RFC 4231 {$label}", $got === $tc['hmac_sha256']);
}

$nn = $k['npamp_inputs'];
$e = $k['expected'];

// --- ORACLE: independent hash_hmac reproduces verify_data. ---
check('finished oracle: verify_data_server',
    bin2hex(hmacOracle(hx($nn['finished_key_server']), hx($nn['th_scv']))) === $e['verify_data_server']);
check('finished oracle: verify_data_client',
    bin2hex(hmacOracle(hx($nn['finished_key_client']), hx($nn['th_ccv']))) === $e['verify_data_client']);

// --- IMPL: computeFinished + verifyFinished accept/reject. ---
$cases = [
    ['server', $nn['finished_key_server'], $nn['th_scv'], $e['verify_data_server']],
    ['client', $nn['finished_key_client'], $nn['th_ccv'], $e['verify_data_client']],
];
foreach ($cases as [$name, $fkHex, $thHex, $want]) {
    $fk = hx($fkHex);
    $th = hx($thHex);
    $wantB = hx($want);

    check("finished impl: computeFinished {$name}",
        bin2hex(Handshake::computeFinished($fk, $th, true)) === $want);
    check("finished impl: verifyFinished accepts correct {$name}",
        Handshake::verifyFinished($fk, $th, $wantB, true));

    $bad = $wantB;
    $bad[0] = chr(ord($bad[0]) ^ 0x01);
    check("finished impl: verifyFinished rejects tampered {$name}",
        !Handshake::verifyFinished($fk, $th, $bad, true));
}

echo $failures === 0 ? "ALL PASS\n" : "FAILURES: {$failures}\n";
exit($failures === 0 ? 0 : 1);
